More detailed explanation on how to set up ``pyenv`` on ``Linux`` and… (#367)
EnergySystemsModellingLab/MUSE_OS: docs/installation/pipx-based.rst

    The next sections will explain in detail the appropriate steps as well as commenting on possible caveats. **We strongly encourage you to read through the sections below** to understand what these steps entitle, but in the end, what we are going to do to install MUSE is the following:

    - Open a terminal

>     - Install `pyenv <https://github.com/pyenv/pyenv>`_ (Linux and MacOS) or `pyenv-win <https://pyenv-win.github.io/pyenv-win/>`_ (Windows) and ascertain it works by invoking `pyenv --version` in the terminal.

Reviewer: Let's keep the tone informal:

```suggestion
    - Install `pyenv <https://github.com/pyenv/pyenv>`_ (Linux and MacOS) or `pyenv-win <https://pyenv-win.github.io/pyenv-win/>`_ (Windows) and make sure it works by invoking `pyenv --version` in the terminal.
```

            python -m pipx ensurepath
            python -m pipx install muse-os

>     - After this, MUSE will be available to use system wide simply by invoking ``muse`` in the terminal. To illustrate this and to finish installation, run ``muse --model default``.

Reviewer: Running the default model isn't really completing the installation. Remember, if users are using `pipx` they are just *running* MUSE, not *developing* it, so won't be running `pytest`.

Author: I know... I wanted to point out to a user to run this command as to stay out of trouble, should they become a developer and run the tests then. Or a potential developer reading this part of the documentation without running a default model.

Maybe I can just say something like:
- After this, MUSE will be available to use system wide simply by invoking ``muse`` in the terminal. To illustrate this and to test your installation, run ``muse --model default``. You should then see a a list of outputs printed to the terminal showing the computations going on in the background.

Alternatively, I can remove line 31 altogether and hope the respective users read the respective parts of the documentation as per intended use.

Reviewer: This is much better. It's a good idea for users to check that things are working -- it was just the "finish the installation" bit that didn't sound quite right.

.. _launch-terminal:

Launching a terminal
~~~~~~~~~~~~~~~~~~~~

> All operating systems have a Terminal application that let you run commands. You will need to use it extensively when using MUSE, so we strongly suggest you get familiar with it. For now, let us just figure out how to launch one:

Reviewer: ```suggestion
All operating systems have a Terminal application that let you run commands. You will need to use it extensively when using MUSE, so we strongly suggest you get familiar with it. For now, let's just figure out how to launch one:
```

Launching a terminal
~~~~~~~~~~~~~~~~~~~~

> All operating systems have a Terminal application that let you run commands. You will need to use it extensively when using MUSE, so we strongly suggest you get familiar with it. For now, let us just figure out how to launch one:

Reviewer: I'd leave this as is. It's a minor point, but "let us" sounds overly formal. We generally try to strike a friendly, informal tone in documentation

To install ``pyenv``, follow these steps:

- **Linux**: In this case, you will need to clone the GitHub repository using ``git``. Most Linux distributions come with ``git`` installed, so this should work out of the box.
Then, complete the setup by adding ``pyenv`` to your profile, so the executable can be found. You can `check the instructions in the official webpage <https://github.com/pyenv/pyenv#set-up-your-shell-environment-for-pyenv>`_,
    or follow the below commands that were tested on `Ubuntu 22.04 LTS` using its popular `bash shell` and `z-shell`. To be specific, we tested them
    on `GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)` and `zsh 5.8.1 (x86_64-ubuntu-linux-gnu)`.

>     We now go through the installation procedure of `pyenv` on Linux, step-by-step:

Reviewer: The extra indentation here is borking the formatting:

![Screenshot from 2024-06-25 at 09_43_16 144810954](https://github.com/EnergySystemsModellingLab/MUSE_OS/assets/23149834/98d725a7-c3f8-40d3-96b8-ebc872d84a83)

>  .. code-block::

Reviewer: It surprised me that there was no `pyenv` package on Ubuntu, but it seems not!

            # Step 1: Install essential libraries needed for pyenv

>             sudo apt install -y make build-essential libssl-dev zlib1g-dev \
>             libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm libncurses5-dev \
>             libncursesw5-dev xz-utils tk-dev libffi-dev liblzma-dev python3-openssl \
>             git
> 
> 

Reviewer: ```suggestion
            sudo apt install -y make build-essential libssl-dev zlib1g-dev \
                libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm libncurses5-dev \
                libncursesw5-dev xz-utils tk-dev libffi-dev liblzma-dev python3-openssl \
                git

```

            # Step 2: Clone the `pyenv` repository for Linux
            git clone https://github.com/pyenv/pyenv.git ~/.pyenv

> 
>             #Step 3: Run one code block in this step only, depending on what shell you use:

Reviewer: ```suggestion
            # Step 3: Run one code block in this step only, depending on what shell you use:
```

    Then, complete the setup by adding ``pyenv`` to your profile, so the executable can be found. `Check the instructions in the official webpage <https://github.com/pyenv/pyenv#set-up-your-shell-environment-for-pyenv>`_.

- **Windows**: ``pyenv-win`` is a separate project but it has the same functionality and it is also simpler to setup.
You can read the detailed installation instructions `from the official pyenv-win website <https://github.com/pyenv-win/pyenv-win/tree/master>`_,
but the easiest way is to run the following command in the ``powershell`` and, upon closing and launching a new shell, you should be ready to go:

    .. code-block:: powershell

>         # Step 1: In your powershell, invoke the following command:
>         `Invoke-WebRequest -UseBasicParsing -Uri "https://raw.githubusercontent.com/pyenv-win/pyenv-win/master/pyenv-win/install-pyenv-win.ps1" -OutFile "./install-pyenv-win.ps1"; &"./install-pyenv-win.ps1"`
> 
>         # Step 2: close the shell you invoked in the command from Step 1 and re-launch an new `powershell`. This is equivalent to invoke the shell command `source` in Linux.
> 
>         # Step 3: Confirm success; you should be returned something similar to `pyenv 3.1.1`
>         `pyenv --version`

Reviewer: A few things:

1.  It's already a code block, so if you put backticks inside it then Sphinx will render them, i.e.:

![Screenshot from 2024-07-02 at 09_14_02 795107487](https://github.com/EnergySystemsModellingLab/MUSE_OS/assets/23149834/26a23bd2-926c-411b-a045-73bb7d4466b6)

2. Relaunching powershell isn't exactly the same as running `source` in bash, so I've just removed that comment

```suggestion
        # Step 1: In your powershell, invoke the following command:
        Invoke-WebRequest -UseBasicParsing -Uri "https://raw.githubusercontent.com/pyenv-win/pyenv-win/master/pyenv-win/install-pyenv-win.ps1" -OutFile "./install-pyenv-win.ps1"; &"./install-pyenv-win.ps1"

        # Step 2: close the shell you invoked in the command from Step 1 and re-launch powershell

        # Step 3: Confirm success; you should be returned something similar to `pyenv 3.1.1`
        pyenv --version
```

Author: Great. I can commit this suggestion. 

Thanks for pointing out that invoking `source` on on the Linux Command Line is not equivalent to re-launching a powershell on Windows. 

I thought about this more deeply and came up with the following:
`source` executes a script within the current shell session, preserving environment variables and other session state changes. Relaunching PowerShell starts a new session, losing any changes made in the previous session. In addition, `source` runs scripts in the same process, while relaunching PowerShell creates a new process.

I won't mention this in the documentation, but are my thoughts correct?

Reviewer: Yeah, exactly :+1:

        Finally open a normal PowerShell and re-run the above installation command.

> Actually installing Python

Reviewer: I know you didn't write this line, but I think this sounds *too* informal!

How about:

```suggestion
Installing your chosen Python version
```

    pyenv install -l

> You should see a long list of versions to choose from. Let's install one of the later version of the 3.9 family:

Reviewer: ```suggestion
You should see a long list of versions to choose from. Let's install one of the later versions of the 3.9 family:
```

    pyenv install -l

> You should see a very long list of versions to choose from. Let us install the latest version of the 3.9 family:

Reviewer: Leave as is

.. code-block:: bash

>     pyenv install 3.9.19

Reviewer: NB: On Linux you can write just `pyenv install 3.9`, but when I last tried this on Windows it didn't work, so I suppose we shouldn't advise users to do it

Installing ``pipx``
~~~~~~~~~~~~~~~~~~~

> Next we need to install ``pipx``, a Python application manager that facilitates installing, keeping applications updated and run them in their own isolated environments.

Reviewer: ```suggestion
Next we need to install ``pipx``, a Python application manager that facilitates installing, keeping applications updated and running them in their own isolated environments.
```

More specifically, ``pipx`` will create a virtual environment to run the tools it installs based on the python version that was used to install pipx to start with, unless you specify another version and that other version is system wide available.

> We could skip this step and install MUSE directly, but that will risk to have conflicting dependencies in the future if you install any other application, breaking your MUSE installation, and we do not want that to happen.

Reviewer: I'm not sure this is quite right. You can point `pipx` at a different Python executable and it'll use it, but it doesn't have to be available systemwide and the way it's written makes it sound like you can just tell it what version of Python to use -- which would be cool, but it doesn't do that.

I'd just drop these two sentences. I think the second one is confusing too.

Reviewer 2: No @alexdewar, @HarmonicReflux is correct. You can tell `pipx` which python version to use when installing a package... as long as that python version is already available. So, if you do: `pipx install muse_os --python=/bin/python3.10`, it will install `muse` in a python 3.10 environment as long as you indicated the path to the existing python3.10 executable, regardless of how `pipx` was installed.

Reviewer 2: If you have `pipx` installed, run `pipx install -h` and check the options.

Reviewer: Right, but that `python3.10` *doesn't* have to be available systemwide! It can be installed wherever (e.g. `~/my_programs/my_python`).

I more meant that you have to tell `pipx` the path to a Python executable rather than just saying "please install using Python 3.7", but I didn't word it particularly well :upside_down_face: -- I know that's not what @HarmonicReflux wrote but I just thought it was a bit ambiguous. But anyway, if users have installed `pipx` via `pyenv` it'll already be using right version of Python, right? So I don't think we want to confuse things by telling users about this option.

Sorry to be pedantic. It's not the most important thing and it's ok as is. I don't want to drag this out for the sake of it...
